### PR TITLE
support dynamic sequence length

### PR DESCRIPTION
**Pull request by RezaYazdaniAminabadi**
I did some changes to the transformer kernel code to support sequence-lengths dynamically.

### CODE REVIEW

**Comment by conglongli** [comment]
Thanks Reza for implementing this feature so fast! Some background motivation: Since bing bert validation data has 512 sequence length, we couldn't calculate the validation loss during seq128 pretraining if we use deepspeed transformer kernel. This is because deepspeed transformer kernel at initialization remember a fixed sequence length. Reza, Elton and I discussed and thought that supporting dynamic sequence length would be quite useful not only for my experiments, and Reza implemented this feature. Reza included a unit test. I will test Reza's implementation in my pre-training experiments and let you know whether it works.

**Comment by RezaYazdaniAminabadi** [comment]
Thanks Conglong for notifying this nice feature :-) I hope it unblocks your testing and we can add this feature to the kernel.

File: deepspeed/pt/deepspeed_cuda.py [hunk 1]
```diff
@@ -174,26 +174,30 @@ def forward(ctx,
...
-                                                   norm_b,
-                                                   config.training,
-                                                   config.pre_layer_norm,
-                                                   config.attn_dropout_checkpoint,
-                                                   config.normalize_invertible,
-                                                   config.gelu_checkpoint)
+         layer_output_dropout_mask,
+         norm2_var,
+         norm2_mean,
+         norm3_var,
```
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply deepspeed/pt/deepspeed_cuda.py:180]: great point! :-) I was about to do that a long time ago and I always forgot!

File: csrc/includes/ds_transformer_cuda.h [hunk 2]
```diff
@@ -121,11 +121,17 @@ class BertTransformerLayer {
...
+                                T*,
+                                T*,
+                                T*,
+                                T*);
 
     inline int GetBatchSize() const { return _batch_size; }
     inline int GetNumHeads() const { return _heads; }
     inline int GetSeqLength() const { return _seq_length; }
+
+    void SetSeqLength(int seq_len, int bsz);
```
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply csrc/includes/ds_transformer_cuda.h:134]: yes it is used in ds_transformer_cuda.cpp: https://github.com/microsoft/DeepSpeed/blob/reyazda/support_dynamic_seqlength/csrc/transformer/ds_transformer_cuda.cpp#L574

File: csrc/includes/dropout.h [hunk 3]
```diff
@@ -18,6 +18,7 @@ class Dropout {
         }
 
         float RATIO() const { return training ? ratio : 0.0; }
+        inline void SetDim(uint32_t d) { dim = d; }
```
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply csrc/includes/dropout.h:21]: yes I remove that

File: csrc/includes/normalize_layer.h [hunk 4]
```diff
@@ -195,33 +157,41 @@ class Normalize_Layer {
...
 
+    inline void SetVar(T* variance)
+    {
+        if (!variance) { throw std::runtime_error("Normalize variance is null."); }
+        vars = variance;
+    }
+
+    inline void SetMean(T* mean)
+    {
+        if (!mean) { throw std::runtime_error("Normalize mean is null."); }
```
[the remark this answers is not in the document]
Comment (RezaYazdaniAminabadi) [reply csrc/includes/normalize_layer.h:188]: No, I cannot do that. These two mean differently, when the use_mean might be true or false based on the layer_norm inversion. However, here this just checks whether the mean is allocated from the outside. So, the SetMean function should not be called when we have normalize_invetible flag to true.

**Comment by HFadeel** [comment]
It will be great if this could be merged. I 'm really excited about it

**Comment by RezaYazdaniAminabadi** [comment]
> It will be great if this could be merged. I 'm really excited about it

**Comment by RezaYazdaniAminabadi** [comment]
Hi @HFadeel ,

We are working on merging this soon. 
Thanks for the interest :)

Reza

**Comment by RezaYazdaniAminabadi** [comment]
Hi @HFadeel 

We have merged these changes into master branch now. Please check in the new changes for DeepSpeed to use the feature. 

Thanks you.
Reza
